### ISSUE TITLE
Disproportionate/unexpected number of poly-T-primed reads mapping to the wrong species

### ISSUE THREAD

**Issue by hepcat72** [hepcat72/split-seq-wrappers#4]
**Problem**

A disproportionate/unexpected number of poly-T-primed reads are mapping to the wrong species.

**Steps to reproduce**

1. Count the number of unique UMIs for each cell barcode combination mapping to each species in exonic and non-exonic locations.

**Current behavior**

When inspecting the "10%" sample (not the "TT" sample), the percentage of poly-T primed reads map to the wrong species at a volume of 24% (non-exonic) and 13% (exonic) for some cell barcode combinations.

**Expected behavior**

The number of poly-t primed reads for each cell barcode should be 2 or 3% of the total, regardless of exonic/non-exonic.

**Suggested Change**

Run a pure mouse sample (mouse brain from the published dataset) against a combined human & mouse reference to see if we get a similarly high percentage of reads mapping to human non-exonic positions.

This is calculated by processing uniquely mapped reads in the split-seq bam file using the code Rob wrote on 3/6/2020, which counts unique UMIs of reads that mapped, for each associated cell barcode.

**Comment**

The data I produced before (on 3/6/2020, according to my notes), which Wei requested be done in another context does not include exon/non-exon in the accounting.

These were the steps I took for the 2 samples (10p and tt).  I did the 10p sample first in 2 separate steps:

```
# Count reads with each UMI

samtools view /Users/rleach/PROJECT/SPLITSEQ/REALTESTRUN/splt10p_single_cells_barcoded_headAligned.sorted.bam | perl -e 'while(<STDIN>){if(/^(.{8})(.{8})(.{8})_(.{10})[^\t]+\t\S+\t([^_]+)\S+\t\S+\t255/){print "$1\t$2\t$3\t$5\t$4\n"}}' | sort | uniq -c > split10p_counts_with_umi.txt

# Count the number of UMI's per cell barcode (and mapped species):

perl -e '$l="";while(<>){s/^ *(\d+) /$1\t/;chomp;@d=split(/\t/);$k="$d[1]\t$d[2]\t$d[3]\t$d[4]";if($l ne $k && $l ne ""){print("$l\t$t\t$c\n");$c=0;$t=0;}$t+=$d[0];$c++;$l=$k}print("$l\t$t\t$c\n");' < split10p_counts_with_umi.txt > split10p_counts_with_umi_counts.txt
```

I then combined those 2 steps into 1 command for the tt sample:

```
samtools view /Users/rleach/PROJECT/SPLITSEQ/REALTESTRUN/splittt_single_cells_barcoded_headAligned.sorted.bam | perl -e 'while(<STDIN>){if(/^(.{8})(.{8})(.{8})_(.{10})[^\t]+\t\S+\t([^_]+)\S+\t\S+\t255/){print "$1\t$2\t$3\t$5\t$4\n"}}' | sort | uniq -c | perl -e '$l="";while(<>){s/^ *(\d+) /$1\t/;chomp;@d=split(/\t/);$k="$d[1]\t$d[2]\t$d[3]\t$d[4]";if($l ne $k && $l ne ""){print("$l\t$t\t$c\n");$c=0;$t=0;}$t+=$d[0];$c++;$l=$k}print("$l\t$t\t$c\n");' > splittt_counts_with_umi_counts.txt
```

**Comment by hepcat72** [comment]
Comment from Wei:

> 1. The disproportionate number of poly-T-primed reads mapped to the wrong species also happens in exonic UMIs (the columns with "True" in your UMI_count table) at 13% (human/mouse), while it is much more frequent in non-exonic UMIs (the columns with "False" in your UMI_count table) at 24%. This analysis can be applied only in the 10PCT sample, not the TT sample.

**Comment by hepcat72** [comment]
I have edited the issue title and description to reflect Wei's comment.

**Comment by hepcat72** [comment]
# ToDo

- [x] Download a pure mouse brain sample from the paper
- [x] Run the pure mouse brain sample against a combined human & mouse reference
- [x] Run the analysis from the comments in the issue on the produced bam file
- [x] Run the conversion of the `*filtered` split-seq outputs to TSV which identifies exonic/non-exonic UMI counts (see code below)

Conversion of final filtered split-seq outputs:
```
foreach d ( *filtered )
cd $d
perl -e 'open(CELLS,$ARGV[0]);my $cn=0;$cellbcs={};while(<CELLS>){@cd=split(/,/,$_);$cd[0]=~s/_/-/;$cellbcs->{$cn}=$cd[0];$cn++;}close(CELLS);open(GTF,$ARGV[3]);$gchrs={};while(<GTF>){if(/^(\S+)\t.*?gene_id[\s"]+([^"]+)/){$chr=$1;$acc=$2;$chr=~s/_/-/g;$gchrs->{$acc}=$chr;}}close(GTF);open(GENES,$ARGV[2]);while(<GENES>){chomp;($idx,$acc,$name,$spc)=split(/,|\t/,$_);$idx++;$genes->{$idx}=(exists($gchrs->{$acc}) ? $gchrs->{$acc} : $spc).".${acc}.$name";}$table=[];$l=0;open(MTX,$ARGV[1]);while(<MTX>){next if(/\%/);$l++;chomp;($cellidx,$geneidx,$count)=split(/\s+/,$_);if($l == 1){$len=$cellidx;print(join("\t",map {exists($cellbcs->{$_}) ? $cellbcs->{$_} : $_} (1..$len)),"\n");next;}$mtx->[$geneidx-1]->[$cellidx-1]=$count}my $i=1;foreach my $row (map {defined($_) ? $_ : [map {0} (1..$len)]} @$mtx){unless(exists($genes->{$i})){print STDERR "ERROR: Gene not found: $i\n"}printrow($row,$len,$genes->{$i});$i++}sub printrow {my $ary = $_[0];$size=$_[1];my $gn=$_[2];print("$gn\t",join("\t",map {defined($ary->[$_]) ? exp2int($ary->[$_]) : "0"} (0..($size-1))),"\n")}sub exp2int {$str=$_[0];if($str =~ /^(.*)e\+(\d+)$/){my $n = $1 * 10 ** $2;$n}else{$str}}' cell_metadata.csv DGE.mtx genes.csv ../../../WEI-SPLITSEQ-TT/hg38_mm10/genes.gtf > DGE.fix.tsv
cd ..
end
```

**Comment by hepcat72** [comment]
So I confirmed that I previously downloaded the mouse brain samples on the cluster here and copied them here:

`/Genomics/biocomp/rleach/SPLITSEQ/mousebrainpaper`

I had used the following commands for the download (using the accessions from the paper):

```
foreach a ( SRR6750056 SRR6750057 SRR6750058 SRR6750059 )
echo Starting $a
fastq-dump $a --split-files
end
```

And in my previous processing, I had been concerned about which wells to specify to extract the specific samples.  That issue was [explained here](https://github.com/yjzhang/split-seq-pipeline/issues/16#issuecomment-568203095) in my interactions with the repo developers.

I should be able to simply run my analysis scripts on the pre-generated output.  Well, actually, I should re-run with the code that has the bug fix and then do that.
